File: cli/runtime.mjs
import { execFile } from 'node:child_process';
import { promisify } from 'node:util';
import { promises as fs } from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { fileURLToPath } from 'node:url';

const execute = promisify(execFile);
const helper = fileURLToPath(new URL('./runtime.exs', import.meta.url));
export function unavailableRuntime(node, reason) {
  return { schemaVersion: 1, status: 'unavailable', node: node || null, collectedAt: new Date().toISOString(), applications: [], processes: [], truncated: false, warnings: [reason] };
}

export async function collectRuntimeSnapshot(root, options = {}) {
  const node = options.node;
  if (!node) return unavailableRuntime(null, 'Connect a local named BEAM node to explore running processes.');
  const match = /^([a-zA-Z0-9_.-]+)@([a-zA-Z0-9_.-]+)$/.exec(node);
  const localHosts = new Set(['localhost', '127.0.0.1', os.hostname(), os.hostname().split('.')[0]]);
  if (!match || !localHosts.has(match[2])) return unavailableRuntime(node, 'Use a node name on this machine (name@hostname).');
  const names = options.nameType || (match[2].includes('.') ? 'longnames' : 'shortnames');
  if (!['longnames', 'shortnames'].includes(names)) return unavailableRuntime(node, 'Node name type must be shortnames or longnames.');
  const limit = Math.max(1, Math.min(10000, Number(options.maxProcesses) || 2000));
  const depth = Math.max(1, Math.min(100, Number(options.maxDepth) || 20));
  try {
    const env = { ...process.env, CODEFLOW_RUNTIME_NODE: node, CODEFLOW_RUNTIME_NAMES: names,
      CODEFLOW_RUNTIME_LIMIT: String(Math.floor(limit)), CODEFLOW_RUNTIME_DEPTH: String(Math.floor(depth)) };
    const cookie = options.cookie ?? process.env.CODEFLOW_BEAM_COOKIE;
    delete env.CODEFLOW_RUNTIME_COOKIE;
    if (cookie) env.CODEFLOW_RUNTIME_COOKIE = cookie;
    const { stdout } = await (options.execute || execute)('elixir', [helper], {
      cwd: root, env, timeout: options.timeout ?? 15000, maxBuffer: 8 * 1024 * 1024,
      killSignal: 'SIGKILL', windowsHide: true
    });
    const line = stdout.split('\n').find(line => line.startsWith('CODEFLOW_RUNTIME:'));
    const raw = JSON.parse(line?.slice('CODEFLOW_RUNTIME:'.length));
    if (!Array.isArray(raw.applications) || !Array.isArray(raw.processes)) throw new Error('Invalid runtime response');
    const canonicalRoot = await fs.realpath(root);
    for (const process of raw.processes) {
      process.sourcePath = null;
      // Compile metadata from a release may refer to another checkout. Never guess.
      if (typeof process.source === 'string' && path.isAbsolute(process.source)) {
        try {
          const source = await fs.realpath(process.source);
          const relative = path.relative(canonicalRoot, source);
          if (relative && !relative.startsWith('..' + path.sep) && relative !== '..' && !path.isAbsolute(relative) && (await fs.stat(source)).isFile()) process.sourcePath = relative.split(path.sep).join('/');
        } catch { /* Source unavailable in this checkout. */ }
      }
      delete process.source;
    }
    return { schemaVersion: 1, status: 'ready', node, collectedAt: new Date().toISOString(), ...raw,
      warnings: [
        ...(raw.processInventoryStatus === 'unavailable' ? ['The process inventory could not be queried; only observed application trees are shown.'] : []),
        ...(raw.truncated ? ['The runtime snapshot reached its process or depth limit.'] : []),
        ...raw.processes.filter(p => p.observationStatus === 'unavailable').map(p => `Process information unavailable for ${p.pid || p.id}; liveness is unknown.`),
        ...(raw.processes.some(p => p.childrenStatus === 'unavailable') ? ['Some supervisors did not respond during this snapshot.'] : [])
      ] };
  } catch {
    // Distribution stderr may contain configuration details: never return it or cookie values.
    return unavailableRuntime(node, 'Could not inspect this node. Check that it is running locally, its name and cookie match, and Elixir 1.19+ is available.');
  }
}

File: cli/runtime.ex
# A disposable distribution client. All target operations are OTP read-only queries.
defmodule Codeflow.Runtime do
  def rpc(node, mod, fun, args), do: :rpc.call(node, mod, fun, args, 1_000)

  def source(node, module) when is_atom(module) and not is_nil(module) do
    case rpc(node, module, :module_info, [:compile]) do
      info when is_list(info) ->
        case Keyword.get(info, :source) do
          source when is_list(source) -> List.to_string(source)
          source when is_binary(source) -> source
          _ -> nil
        end
      _ -> nil
    end
  end
  def source(_, _), do: nil

  def walk(node, pid, parent, application, kind, modules, depth, state, limit, max_depth) do
    key = inspect(pid)
    cond do
      Map.has_key?(state.processes, key) -> state
      map_size(state.processes) >= limit -> %{state | truncated: true}
      true ->
        info = rpc(node, :erlang, :process_info, [pid, [:registered_name, :initial_call, :memory, :message_queue_len, :reductions, :status]])
        # Only :undefined confirms that this PID has exited. An RPC failure
        # (including timeout or node loss) cannot establish process liveness.
        {info, observation_status} = case info do
          info when is_list(info) -> {info, "ready"}
          :undefined -> {[], "exited"}
          _ -> {[], "unavailable"}
        end
        module = case modules do
          [m | _] when is_atom(m) -> m
          _ -> case Keyword.get(info, :initial_call) do
            {m, _, _} when m not in [:proc_lib, :erlang] -> m
            _ -> nil
          end
        end
        {source_path, sources} = case Map.fetch(state.sources, module) do
          {:ok, path} -> {path, state.sources}
          :error -> path = source(node, module); {path, Map.put(state.sources, module, path)}
        end
        name = Keyword.get(info, :registered_name)
        label = if is_atom(name) and not is_nil(name), do: Atom.to_string(name), else: key
        child_result = if kind == :supervisor and depth < max_depth do
          case rpc(node, :supervisor, :which_children, [pid]) do
            children when is_list(children) -> {:ok, Enum.filter(children, fn {_, child, _, _} -> is_pid(child) end)}
            _ -> {:unavailable, []}
          end
        else
          {if(kind == :supervisor, do: :limited, else: :ok), []}
        end
        {children_status, children} = child_result
        entry = %{id: key, pid: key, parentId: parent, application: application, label: label,
          observationStatus: observation_status,
          type: Atom.to_string(kind), module: if(module, do: Atom.to_string(module), else: nil),
          source: source_path, childrenStatus: Atom.to_string(children_status), children: Enum.map(children, fn {_, child, _, _} -> inspect(child) end),
          metrics: %{memory: Keyword.get(info, :memory), messageQueueLength: Keyword.get(info, :message_queue_len),
            reductions: Keyword.get(info, :reductions), status: to_string(Keyword.get(info, :status, observation_status))}}
        state = %{state | processes: Map.put(state.processes, key, entry), sources: sources,
          truncated: state.truncated or (kind == :supervisor and depth >= max_depth)}
        Enum.reduce(children, state, fn {_, child, type, mods}, acc ->
          walk(node, child, key, application, type, mods, depth + 1, acc, limit, max_depth)
        end)
    end
  end

  def process_inventory(node) do
    case rpc(node, :erlang, :processes, []) do
      pids when is_list(pids) -> {"ready", pids}
      _ -> {"unavailable", []}
    end
  end

  def run do
    target = System.fetch_env!("CODEFLOW_RUNTIME_NODE") |> String.to_atom()
    mode = if System.get_env("CODEFLOW_RUNTIME_NAMES") == "longnames", do: :longnames, else: :shortnames
    [_name, host] = String.split(Atom.to_string(target), "@", parts: 2)
    own = String.to_atom("codeflow_#{System.pid()}_#{System.unique_integer([:positive])}@#{host}")
    {:ok, _} = Node.start(own, mode)
    if cookie = System.get_env("CODEFLOW_RUNTIME_COOKIE"), do: Node.set_cookie(String.to_atom(cookie))
    if Node.connect(target) != true, do: raise("Target node unavailable")
    limit = System.fetch_env!("CODEFLOW_RUNTIME_LIMIT") |> String.to_integer()
    max_depth = System.fetch_env!("CODEFLOW_RUNTIME_DEPTH") |> String.to_integer()
    apps = case rpc(target, :application, :which_applications, []) do
      apps when is_list(apps) -> Enum.sort(apps)
      _ -> raise("Applications unavailable")
    end
    initial = %{processes: %{}, sources: %{}, truncated: false}
    {applications, state} = Enum.map_reduce(apps, initial, fn {app, description, version}, state ->
      master = rpc(target, :application_controller, :get_master, [app])
      root = if is_pid(master), do: rpc(target, :application_master, :get_child, [master]), else: nil
      {pid, module} = case root do
        {pid, module} when is_pid(pid) -> {pid, module}
        _ -> {nil, nil}
      end
      state = if pid, do: walk(target, pid, nil, Atom.to_string(app), :supervisor, [module], 0, state, limit, max_depth), else: state
      {%{name: Atom.to_string(app), description: to_string(description), version: to_string(version), rootId: if(pid, do: inspect(pid), else: nil)}, state}
    end)
    # Include unsupervised processes too; their callback identity may be unknown.
    {inventory_status, pids} = process_inventory(target)
    state = Enum.reduce(pids, state, fn pid, state -> walk(target, pid, nil, nil, :process, [], 0, state, limit, max_depth) end)
    processes = Map.values(state.processes) |> Enum.sort_by(& &1.id)
    IO.puts("CODEFLOW_RUNTIME:" <> JSON.encode!(%{applications: applications, processes: processes, processInventoryStatus: inventory_status, truncated: state.truncated}))
  end
end
